src/utils/decrypt-data.ts

```diff
@@ -0,0 +1,95 @@
+import {
+  EncryptionContext,
+  KmsKeyringNode,
+  buildDecrypt,
+} from "@aws-crypto/client-node";
+import buildKmsKeyring from "./kms-keyring-builder";
+import { logger } from "./logger";
+import { getHashedAccessCheckValue } from "./get-access-check-value";
+
+const MAX_ENCRYPTED_DATA_KEY = 5;
+const DECODING = "utf8";
+const ENCODING = "base64";
+
+const decryptClientConfig = { maxEncryptedDataKeys: MAX_ENCRYPTED_DATA_KEY };
+const decryptClient = buildDecrypt(decryptClientConfig);
+
+let keyring: KmsKeyringNode;
+
+export async function generateExpectedContext(
+  userId: string
+): Promise<EncryptionContext> {
+  const { AWS_REGION, ACCOUNT_ID, ENVIRONMENT, VERIFY_ACCESS_VALUE } =
+    process.env;
+  if (AWS_REGION === undefined) {
+    logger.error("Missing AWS_REGION environment variable");
+  }
+
+  if (ACCOUNT_ID === undefined) {
+    logger.error("Missing ACCOUNT_ID environment variable");
+  }
+
+  if (ENVIRONMENT === undefined) {
+    logger.error("Missing ENVIRONMENT environment variable");
+  }
+
+  if (VERIFY_ACCESS_VALUE === undefined) {
+    logger.error("Missing VERIFY_ACCESS_VALUE environment variable");
+  }
+
+  let accessCheckValue;
+  try {
+    accessCheckValue = await getHashedAccessCheckValue(VERIFY_ACCESS_VALUE);
+  } catch (error) {
+    logger.error("Unable to obtain Access Verification value.");
+  }
+
+  return {
+    origin: AWS_REGION,
+    accountId: ACCOUNT_ID,
+    stage: ENVIRONMENT,
+    userId,
+    accessCheckValue,
+  };
+}
+
+export function validateEncryptionContext(
+  context: EncryptionContext,
+  expected: EncryptionContext
+): void {
+  if (context === undefined || Object.keys(context).length === 0) {
+    logger.error("Encryption context is empty or undefined");
+  }
+
+  Object.keys(expected).forEach((key) => {
+    if (context[key] !== expected[key]) {
+      logger.error(`Encryption context mismatch: ${key}`);
+    }
+  });
+}
+
+export async function decryptData(
+  data: string,
+  userId: string
+): Promise<string> {
+  try {
+    logger.debug({
+      trace: "decryptData",
+      message: "uid",
+      userid: userId,
+    });
+    keyring ??= await buildKmsKeyring();
+    const result = await decryptClient.decrypt(
+      keyring,
+      Buffer.from(data, ENCODING)
+    );
+    validateEncryptionContext(
+      result.messageHeader.encryptionContext,
+      await generateExpectedContext(userId)
+    );
+    return result.plaintext.toString(DECODING);
+  } catch (error) {
+    logger.error(error, "Failed to decrypt data.");
```

Comment:
Should we include the trace id in this log message so we can determine which user it impacted?

Reply:
I think it's useful yes. I'll add that

Reply:
Done